runtime/onert/core/src/ir/train/TrainableGraph.cc

```diff
@@ -229,6 +230,46 @@ std::vector<ir::OperationIndex> TrainableGraph::btopolSortOperations() const
   return ret;
 }
 
+std::vector<ir::OperationIndex>
+TrainableGraph::truncateBackwardOrder(std::vector<ir::OperationIndex> backward_order) const
+{
+  auto forward_order = backward_order;
+  std::reverse(forward_order.begin(), forward_order.end());
+
+  std::set<ir::OperationIndex> alive;
+
+  for (const auto &index : forward_order)
+  {
+    const auto &op = operations().at(index);
+    const auto &trainable_op = dynamic_cast<const ITrainableOperation &>(op);
+
+    if (trainable_op.hasTrainableParameter())
+      alive.insert(index);
```

Comment:
At this point, `index` is in `alive` if and only if it either has a trainable parameter, or it is reachable by a node with a trainable parameter.